p2p/src/peer.rs

```diff
@@ -0,0 +1,179 @@
+use std::collections::HashMap;
+use std::thread;
+
+use eyre::Result;
+use flume::{self, Receiver, Sender};
+
+use tendermint::node;
+use tendermint::public_key::PublicKey;
+
+use crate::message;
+use crate::transport::{Connection, Direction, StreamId};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {}
+
+pub trait State: private::Sealed {}
+
+pub struct Connected<Conn> {
+    connection: Direction<Conn>,
+}
+impl<Conn> State for Connected<Conn> {}
+
+pub struct Running<Conn> {
+    connection: Direction<Conn>,
+    pub receiver: Receiver<message::Receive>,
+    senders: HashMap<StreamId, Sender<Box<dyn message::Outgoing>>>,
+}
+impl<Conn> State for Running<Conn> {}
+
+pub struct Stopped {
+    error: Option<Error>,
+}
+impl State for Stopped {}
+
+pub struct Peer<St> {
+    pub id: node::Id,
+
+    pub state: St,
+}
+
+impl<Conn> From<Direction<Conn>> for Peer<Connected<Conn>>
+where
+    Conn: Connection,
+{
+    fn from(connection: Direction<Conn>) -> Peer<Connected<Conn>> {
+        let pk = match &connection {
+            Direction::Incoming(conn) => conn.public_key(),
+            Direction::Outgoing(conn) => conn.public_key(),
+        };
+
+        let id = match pk {
+            PublicKey::Ed25519(ed25519) => node::Id::from(ed25519),
+            _ => panic!(),
+        };
+
+        Peer {
+            id,
+
+            state: Connected { connection },
+        }
+    }
+}
+
+impl<Conn> Peer<Connected<Conn>>
+where
+    Conn: Connection,
+{
+    pub fn run(self, stream_ids: Vec<StreamId>) -> Result<Peer<Running<Conn>>> {
+        // FIXME(xla): No queue should be unbounded, backpressure should be finley tuned and/or
```

Comment:
is there a way for `Peer` to signal `Supervisor` that it must stop reading bytes from `Connection` (more than the peer can handle at the moment)?

Reply:
Could you explain how the peer knows it cant handle more? As in how would it determine that?

Reply:
because each peer has a queue (which at the moment is unbounded, but I suppose will be of a fixed size)

Reply:
Good to raise this early. Intuitively this would necessitate to make backpressue a first-class concept, which protocols that rely on it have to handle. Particularly if a send to a peer fails it should generate a new event which is passed back to the rest of the system. Which in turn could be used to decide if certain/all messages are being held or dropped. A major challenge here is that the message that experiences backpressure might not be the offending one as the message streams to peers are heterogenous. It could be more specific as in which one of the streams inside of a peer is slow to push to. Open to suggestions of how to incorporate it in the rest of the system.